063-topic-replication-factor-change.md

```diff
@@ -0,0 +1,271 @@
+| STRIMZI_CRUISE_CONTROL_RACK_ENABLED    | Boolean | false   | Whether rack awareness is enabled in the target Kafka cluster |
+| STRIMZI_CRUISE_CONTROL_SECRET_NAME     | String  | ""      | Secret name containing the truststore                         |
+| STRIMZI_CRUISE_CONTROL_API_SECRET_NAME | String  | ""      | Secret name containing the API credentials                    |
+
+With Cluster Operator deployment, a dedicated admin user is automatically created for the Topic Operator.
+With standalone deployment, the Cluster Operator expects two secrets that needs to be created by the user:
+
+- A secret with `cruise-control.p12` key containing the truststore in PKCS12 format, and `cruise-control.password` key containing the truststore password.
+- A secret with `cruise-control.apiToAdminUsername` key containing the API username with admin role, and `cruise-control.apiToAdminPassword` key containing the user password.
+
+### KafkaTopic CRD changes
+
+During the whole replicas change process the KafkaTopic remains in a Ready state.
+A replicas change is executed asynchronously, and takes more than one reconciliation to complete.
+
+A new optional KafkaTopicStatus subsection called `replicasChange` is used to update the user, and keep track of the replicas change state.
+At any given time, the replicas change can be pending, ongoing, or completed.
+
+- **Pending**: Not in Cruise Control's task queue (not yet sent or request error).
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:50.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 2
+    replicasChange:
+      completed: false
+    topicName: my-topic
+  ```
+
+- **Ongoing**: In Cruise Control's task queue, but execution not started, or not completed.
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:53.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 3
+    replicasChange:
+      completed: false
+      userTaskId: 1aa418ca-53ed-4b93-b0a4-58413c4fc0cb
+    topicName: my-topic
+  ```
+
+- **Completed**: Cruise Control's task execution completed (target replication factor reconciled).
+  ```sh
+  status:
+    conditions:
+    - lastTransitionTime: "2024-01-18T16:13:58.490918232Z"
+      status: "True"
+      type: Ready
+    observedGeneration: 4
+    topicName: my-topic
+  ```
+
+### Reconciliation logic
+
+At runtime, the Topic Operator watches and periodically reconciles the configuration of all managed and unpaused KafkaTopic resources.
+Among the other configurations, the reconciliation logic detects `KafkaTopic.spec.replicas` changes by comparing with the topic replication factor value in Kafka.
+
+A ReplicasChangeClient is used to isolate the logic required to call the Cruise Control REST API from the rest of the reconciliation logic.
+The ReplicasChangeClient supports sending multiple replicas changes at the same time.
```

Comment:
Does this mean batching multiple topics into a single request? Or just having separate requests for different topics sent in parallel?

Reply:
I agree with need a bit more clarity on this. I took it to mean multiple topics get batched into a single request. But we need to understand _how_ that happens. Imagine the following scenario:

* User has some script/tooling to update `KafkaTopic` RF in bulk.
* The script does a `kubteclt patch` on the `KafkaTopics` in a loop.
* The TO reacts to those watch events one at a time
* What we want to happen: We hit the `topic_configuration` endpoint once, with all those topics in the request.
* What the easy implementation does: We hit the `topic_configuration` endpoint once for each topic.

We want the former behaviour because it's reduces the amount of time the user has to wait for their changes to take effect, and because it means CC can do a single set of data movements, avoiding unnecessary work. But it's not completely trivial to get this behaviour. One way to do it would be to have a queue of RF changes. The regular reconciliation logic adds to the queue, and we process this queue in batches. But we'd need to define it in more detail how this happens, and what configuration parameters ought to be exposed.

Reply:
I disagree. I think we want to go for the simplicity of the implementation over having an amazing efficiency in niche use-cases. (Also ... LOL ... we have been discussing this for several weeks and Federico just changed this :-o)

Reply:
I switched to batch/group requests, and it's not more complicated, really. I just have to group the topics by RF value and create the json. There is no additional queue, the reconciliation logic already has one, and provides a batch of replicas changes. 

This is useful with bulk replicas change, which can happen in some edge cases. A potentially big backlog of messages may be created when the Topic Operator is not running, or there is an issue with Cruise Control or Kafka. Another example is a GitOps pipeline delivering many replicas changes at the same time to reduce the storage occupation of non-critical topics.

I tried my draft implementation with 100 replicas updates applied at the same time (not in a loop) on topics with some data. This created 2 or 3 Cruise Control's requests and completed in less that 5 minutes. I think this is good enough behavior.

Reply:
I guess this is not obvious from the proposal -> neither that it works like that nor how it is actually implemented.

Reply:
Yeah, I realized that and pushed some changes in the "Topic configuration" section.